// ra-ui-materialui/dist/cjs/list/datatable/DataTableHeadCell.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTableHeadCell = void 0;
var React = __importStar(require("react"));
var ra_core_1 = require("ra-core");
var material_1 = require("@mui/material");
var styles_1 = require("@mui/material/styles");
var clsx_1 = __importDefault(require("clsx"));
var DataTableRoot_1 = require("./DataTableRoot");
var oppositeOrder = {
    ASC: 'DESC',
    DESC: 'ASC',
};
var PREFIX = 'RaDataTableHeadCell';
exports.DataTableHeadCell = React.memo(React.forwardRef(function (inProps, ref) {
    var props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    var cellSx = props.cellSx, className = props.className, cellClassName = props.cellClassName, disableSort = props.disableSort, headerClassName = props.headerClassName, field = props.field, render = props.render, source = props.source, label = props.label, sortByOrder = props.sortByOrder, rest = __rest(props, ["cellSx", "className", "cellClassName", "disableSort", "headerClassName", "field", "render", "source", "label", "sortByOrder"]);
    var _a = (0, ra_core_1.useDataTableStoreContext)(), storeKey = _a.storeKey, defaultHiddenColumns = _a.defaultHiddenColumns;
    var sort = (0, ra_core_1.useDataTableSortContext)();
    var handleSort = (0, ra_core_1.useDataTableCallbacksContext)().handleSort;
    var hiddenColumns = (0, ra_core_1.useStore)(storeKey, defaultHiddenColumns)[0];
    var resource = (0, ra_core_1.useResourceContext)();
    var translate = (0, ra_core_1.useTranslate)();
    var translateLabel = (0, ra_core_1.useTranslateLabel)();
    var isColumnHidden = hiddenColumns.includes(source);
    if (isColumnHidden)
        return null;
    var nextSortOrder = sort && sort.field === source
        ? oppositeOrder[sort.order]
        : sortByOrder !== null && sortByOrder !== void 0 ? sortByOrder : 'ASC';
    var fieldLabel = translateLabel({
        label: typeof label === 'string' ? label : undefined,
        resource: resource,
        source: source,
    });
    var sortLabel = translate('ra.sort.sort_by', {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() +
                fieldLabel.slice(1)
            : undefined,
        order: translate("ra.sort.".concat(nextSortOrder)),
        _: translate('ra.action.sort'),
    });
    return (React.createElement(TableCellStyled, __assign({ ref: ref, className: (0, clsx_1.default)(DataTableRoot_1.DataTableClasses.headerCell, className, headerClassName, "column-".concat(source)), variant: "head" }, rest), handleSort && sort && !disableSort && source ? (React.createElement(material_1.Tooltip, { title: sortLabel, placement: props.align === 'right'
            ? 'bottom-end'
            : 'bottom-start', enterDelay: 300 },
        React.createElement(material_1.TableSortLabel, { active: sort.field === source, direction: sort.order === 'ASC' ? 'asc' : 'desc', "data-field": source, "data-order": sortByOrder || 'ASC', onClick: handleSort },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource })))) : (React.createElement(material_1.TableSortLabel, { disabled: true },
        React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource })))));
}));
exports.DataTableHeadCell.displayName = 'DataTableHeadCell';
var TableCellStyled = (0, styles_1.styled)(material_1.TableCell, {
    name: PREFIX,
    overridesResolver: function (props, styles) { return styles.root; },
})(function () { return ({}); });
